dhis-2/dhis-services/dhis-service-reporting/src/main/java/org/hisp/dhis/predictor/DefaultPredictionService.java

```diff
@@ -323,136 +328,257 @@ public void predict( Predictor predictor, Date startDate, Date endDate, Predicti
+        {
+            periods.addAll( samplePeriods );
+        }
+
+        if ( !outputPeriodItems.isEmpty() )
+        {
+            periods.addAll( outputPeriods );
+        }
+
+        return periods;
+    }
+
+    /**
+     * Categories DimensionalItemObjects found in the predictor expression (and
+     * skip test) according to how their values will be fetched from either the
+     * datavalue table or analytics.
+     *
+     * @param items DimensionalItemObjects to be categorized.
+     * @param dataElements datavalue: data elements.
+     * @param dataElementOperands datavalue: data element operands.
+     * @param analyticsAttributeOptionItems analytics: items stored with AOCs.
+     * @param analyticsNonAttributeOptionItems analytics: items without AOCs.
+     */
+    private void categorizeItems( Set<DimensionalItemObject> items,
+        Set<DataElement> dataElements, Set<DataElementOperand> dataElementOperands,
+        Set<DimensionalItemObject> analyticsAttributeOptionItems,
+        Set<DimensionalItemObject> analyticsNonAttributeOptionItems )
+    {
+        for ( DimensionalItemObject i : items )
+        {
+            if ( i instanceof DataElement )
+            {
+                dataElements.add( (DataElement) i );
+            }
+            else if ( i instanceof DataElementOperand )
+            {
+                dataElementOperands.add( (DataElementOperand) i );
+            }
+            else if ( hasAttributeOptions( i ) )
+            {
+                analyticsAttributeOptionItems.add( i );
+            }
+            else
+            {
+                analyticsNonAttributeOptionItems.add( i );
+            }
+        }
+    }
+
+    /**
+     * Add the non-deleted, numeric values from the datavalue table to the map
+     * of data by attributeOptionCombo.
+     * <p>
+     * The two types of dimensional item object that are needed from the data
+     * value table are DataElement (the sum of all category option combos for
+     * that data element) and DataElementOperand (a particular combination of
+     * DataElement and CategoryOptionCombo).
+     *
+     * @param dataValues List of data values.
+     * @param aocData Map of attributeOptionCombo-keyed data.
+     * @param items the items we will need for expression evaluation.
+     */
+    private void addDataValuesToAocData( List<DataValue> dataValues,
+        MapMapMap<String, Period, DimensionalItemObject, Double> aocData,
+        Set<DimensionalItemObject> items )
+    {
+        for ( DataValue dv : dataValues )
+        {
+            if ( dv.isDeleted() || dv.getValue() == null )
+            {
+                continue;
+            }
+
+            Double value;
+
+            try
+            {
+                value = Double.parseDouble( dv.getValue() );
+            }
+            catch ( NumberFormatException e )
+            {
+                continue; // Ignore any non-numeric values.
+            }
```

Comment:
A way to avoid multiple `continue`s would be to extract this to a helper method. I'd guess we do something similar elsewhere too.

Reply:
Great idea. Done.